src/Compilers/CSharp/Portable/Parser/Lexer.cs

```diff
@@ -921,11 +921,65 @@ private bool ScanInteger()
             return start < TextWindow.Position;
         }
 
+        // Allows underscores in integers, except at beginning and end
+        private void ScanNumericLiteralSingleInteger(StringBuilder builder, ref bool underscoreInWrongPlace, ref bool usedUnderscore, bool isHex, bool isBinary)
+        {
+            if (TextWindow.PeekChar() == '_')
+            {
+                underscoreInWrongPlace = true;
+            }
+
+            char ch;
+            var lastCharWasUnderscore = false;
+            while (true)
```

Comment:
Can we add a test where the text input ends in a binary literal to ensure this loop properly terminates when presented with an end of input situation?